File: 1320_Minimum_Distance_to_Type_a_Word_Using_Two_Fingers/solution.py
class Solution:
    def minimumDistance(self, word: str) -> int:
        dist = lambda x, y: (
            0 if 26 in (x, y) else abs(x // 6 - y // 6) + abs(x % 6 - y % 6)
        )
        INF, n = 1 << 30, len(word)
        dp = [[INF] * 27 for _ in range(n)]
        dp[0][26] = 0
        p = ord(word[0]) - 65
        for i, c in enumerate(word[1:], 1):
            x = ord(c) - 65
            for j in range(27):
                dp[i][j] = min(dp[i][j], dp[i - 1][j] + dist(p, x))
                dp[i][p] = min(dp[i][p], dp[i - 1][j] + dist(j, x))
            p = x
        return min(dp[-1])

File: 1320_Minimum_Distance_to_Type_a_Word_Using_Two_Fingers/solution.c
int gd(int p, int q) { return abs(p/6 - q/6) + abs(p%6 - q%6); }
int minimumDistance(char* w) {
    int n = strlen(w);
    int dp[n][26][26];
    for (int i=0;i<n;i++) for (int j=0;j<26;j++) for (int k=0;k<26;k++)
    dp[i][j][k] = INT_MAX/2;
    for (int i=0;i<26;i++) dp[0][i][w[0]-'A'] = dp[0][w[0]-'A'][i] = 0;
    for (int i=1;i<n;i++) {
        int c=w[i]-'A', p=w[i-1]-'A', d=gd(p,c);
        for (int j=0;j<26;j++) {
            dp[i][c][j] = fmin(dp[i][c][j], dp[i-1][p][j]+d);
            dp[i][j][c] = fmin(dp[i][j][c], dp[i-1][j][p]+d);
            if (p==j) for (int k=0;k<26;k++) {
                int d0=gd(k,c);
                dp[i][c][j] = fmin(dp[i][c][j], dp[i-1][k][j]+d0);
                dp[i][j][c] = fmin(dp[i][j][c], dp[i-1][j][k]+d0);
            }
        }
    }
    int ans=INT_MAX/2;
    for (int i=0;i<26;i++) for (int j=0;j<26;j++) if (dp[n-1][i][j]<ans) ans=dp[n-1][i][j];
    return ans;
}
